### PR TITLE
[SPARK-35180][BUILD] Allow to build SparkR with SBT

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR proposes a change that allows us to build SparkR with SBT.

### Why are the changes needed?

In the current master, SparkR can be built only with Maven.
It's helpful if we can built it with SBT.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

I confirmed that I can build SparkR on Ubuntu 20.04 with the following command.
```
build/sbt -Psparkr package
```

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @srowen and @HyukjinKwon

**Review by dongjoon-hyun — COMMENTED** [review]
It's great. Could you add the build command into `R/README.md` along with the existing `mvn` command, @sarutak ?
- https://github.com/apache/spark/blame/master/R/README.md#L23

**Comment by dongjoon-hyun** [comment]
BTW, I can do the AS-IS master branch, what is the missing before this PR?
```
$ build/sbt package -Psparkr
$ R/install-dev.sh
$ bin/sparkR

R version 3.5.2 (2018-12-20) -- "Eggshell Igloo"
Copyright (C) 2018 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin15.6.0 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

Launching java with spark-submit command /Users/dongjoon/APACHE/spark-merge/bin/spark-submit   "sparkr-shell" /var/folders/p5/tzyffbvj0md4djphs4fym9zw0000gn/T//RtmpV1x8wW/backend_portc786d577092
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
21/04/21 22:02:46 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
21/04/21 22:02:47 WARN SQLUtils: SparkR: enableHiveSupport is requested for SparkSession but Spark is not built with Hive or spark.sql.catalogImplementation is not set to 'hive', falling back to without Hive support.

Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.2.0-SNAPSHOT
      /_/


SparkSession Web UI available at http://localhost:4040
SparkSession available as 'spark'(master = local[*], app id = local-1619067767184).
>
```

**Comment by sarutak** [comment]
> It's great. Could you add the build command into R/README.md along with the existing mvn command, @sarutak ?

O.K, I'll do it.

> BTW, I can do the AS-IS master branch, what is the missing before this PR?

`mvn -Psparkr` runs `R/install-dev.sh` automatically but SBT didn't.

**Comment by dongjoon-hyun** [comment]
Got it!
> mvn -Psparkr runs R/install-dev.sh automatically but SBT didn't.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

File: R/README.md [hunk 1]
```diff
@@ -17,10 +17,14 @@ export R_HOME=/home/username/R
 
 #### Build Spark
 
-Build Spark with [Maven](https://spark.apache.org/docs/latest/building-spark.html#buildmvn) and include the `-Psparkr` profile to build the R package. For example to use the default Hadoop versions you can run
+Build Spark with [Maven](https://spark.apache.org/docs/latest/building-spark.html#buildmvn) or [SBT](http://spark.apache.org/docs/latest/building-spark.html#building-with-sbt), and include the `-Psparkr` profile to build the R package. For example to use the default Hadoop versions you can run
```
Comment (dongjoon-hyun) [inline R/README.md:20]: Oh, `http` -> `https`.

Comment (sarutak) [reply R/README.md:20]: Ah, I just copied from the URL bar. I'll change it. Thanks.

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -888,6 +892,22 @@ object PySparkAssembly {
...
+  import scala.sys.process.Process
+
+  val buildRPackage = taskKey[Unit]("Build the R package")
+  lazy val settings = Seq(
+    buildRPackage := {
+      val command = baseDirectory.value / ".." / "R" / "install-dev.sh"
+      Process(command.toString).!!
+    },
+    (Compile / compile) := {
+      buildRPackage.value
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala:905]: I think we build SparkR after the compilation of core so the order should probably be switched with the next line.

**Review by HyukjinKwon — APPROVED** [review]
LGTM2

File: project/SparkBuild.scala (hunk 2 above)
Comment (sarutak) [reply project/SparkBuild.scala]: Thanks. I've updated.

**Comment by sarutak** [comment]
retest this please.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**